src/main/java/com/limechain/trie/BlockTrieAccessor.java

```diff
@@ -0,0 +1,197 @@
+package com.limechain.trie;
+
+import com.limechain.runtime.version.StateVersion;
+import com.limechain.storage.DeleteByPrefixResult;
+import com.limechain.storage.KVRepository;
+import com.limechain.storage.trie.TrieStorage;
+import com.limechain.trie.structure.Entry;
+import com.limechain.trie.structure.NodeHandle;
+import com.limechain.trie.structure.TrieStructure;
+import com.limechain.trie.structure.Vacant;
+import com.limechain.trie.structure.database.NodeData;
+import com.limechain.trie.structure.nibble.Nibble;
+import com.limechain.trie.structure.nibble.Nibbles;
+import io.emeraldpay.polkaj.types.Hash256;
+import lombok.AllArgsConstructor;
+import org.apache.commons.lang3.NotImplementedException;
+
+import java.util.List;
+import java.util.Optional;
+
+@AllArgsConstructor
+public class BlockTrieAccessor implements KVRepository<Nibbles, byte[]> {
+    public static final String TRANSACTIONS_NOT_SUPPORTED = "Block Trie Accessor does not support transactions.";
+    private final Hash256 blockHash;
+    private final TrieStructure<NodeData> partialTrie;
+    private final TrieStorage trieStorage;
+
+    public BlockTrieAccessor(Hash256 blockHash) {
+        this.blockHash = blockHash;
+        this.trieStorage = TrieStorage.getInstance();
+        this.partialTrie = new TrieStructure<>();
+    }
+
+    @Override
+    public boolean save(Nibbles key, byte[] value) {
+        loadPathToKey(key);
+        loadChildren(key);
+        NodeData nodeData = new NodeData(value);
+        partialTrie.insertNode(key, nodeData);
+        markForRecalculation(key);
+
+        return true;
+    }
+
+    private void markForRecalculation(Nibbles key) {
+        NodeHandle<NodeData> createdNode = partialTrie.node(key).asNodeHandle();
+        Optional.ofNullable(createdNode.getParent())
+                .map(NodeHandle::getUserData)
+                .ifPresent(data -> data.setMerkleValue(null));
+        for (Nibble nibble : Nibbles.ALL) {
+            createdNode.getChild(nibble)
+                    .map(NodeHandle::getUserData)
+                    .ifPresent(data -> data.setMerkleValue(null));
+        }
+    }
+
+    /**
+     * Load the path up to the given key, from the TrieStorage into our partial trie.
+     *
+     * @param key key
+     */
+    private void loadPathToKey(Nibbles key) {
+        Entry<NodeData> closestNode = partialTrie.node(key);
+        if (closestNode instanceof Vacant<NodeData> vacantNode) {
+            Optional.of(vacantNode)
+                    .map(Vacant::getClosestAncestorIndex)
+                    .map(partialTrie::nodeHandleAtIndex)
+                    .map(NodeHandle::getUserData)
+                    .map(NodeData::getMerkleValue)
+                    .map(closestMerkle -> trieStorage.entriesBetween(closestMerkle, key.toString()))
+                    .ifPresent(entries -> entries.forEach(e -> partialTrie.insertNode(e.key(), e.nodeData())));
+        }
+    }
```

Comment:
We never load any nodes to the partialTrie and it is always empty, thus getClosestAncestorIndex returns no value.